File: RBE-3002-D26-Team13-fast_pursuit/exploration/src/novelty_mapping.cpp
#include "novelty_mapping.hpp"

#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/occupancy_grid.hpp>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <stdexcept>
#include <string>

// ---------------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------------
NoveltyMapping::NoveltyMapping()
: Node("Novelty_Node")
{
    cb_group_ = create_callback_group(
        rclcpp::CallbackGroupType::MutuallyExclusive);

    rclcpp::SubscriptionOptions sub_opts;
    sub_opts.callback_group = cb_group_;

    subscriber_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
        "/map", 10,
        std::bind(&NoveltyMapping::mapCallback, this, std::placeholders::_1),
        sub_opts);

    novelty_map_publisher_  = create_publisher<nav_msgs::msg::OccupancyGrid>("/novelty",   10);
    frontier_map_publisher_ = create_publisher<nav_msgs::msg::OccupancyGrid>("/frontier",  10);

    // Pre-compute ray directions
    cos_dirs_.resize(NUM_RAY_CASTS);
    sin_dirs_.resize(NUM_RAY_CASTS);
    for (int i = 0; i < NUM_RAY_CASTS; ++i) {
        double angle = 2.0 * M_PI * i / NUM_RAY_CASTS;
        cos_dirs_[i] = std::cos(angle);
        sin_dirs_[i] = std::sin(angle);
    }
}

// ---------------------------------------------------------------------------
// /map callback
// ---------------------------------------------------------------------------
void NoveltyMapping::mapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr map)
{
    const int height = static_cast<int>(map->info.height);
    const int width  = static_cast<int>(map->info.width);

    // Reshape flat data → 2-D row-major array (row = y, col = x — same as Python)
    map_.assign(height, std::vector<int>(width));
    for (int r = 0; r < height; ++r)
        for (int c = 0; c < width; ++c)
            map_[r][c] = static_cast<int>(map->data[r * width + c]);

    map_info_ = map->info;

    // --- Frontier ---
    frontier_map_ = findFrontier(map_);

    {
        nav_msgs::msg::OccupancyGrid frontier_grid;
        frontier_grid.header = map->header;
        frontier_grid.info   = map->info;
        frontier_grid.data.resize(height * width);
        for (int r = 0; r < height; ++r)
            for (int c = 0; c < width; ++c)
                frontier_grid.data[r * width + c] =
                    static_cast<int8_t>(std::clamp(frontier_map_[r][c], 0, 100));
        frontier_map_publisher_->publish(frontier_grid);
    }

    // --- Novelty ---
    novelty_map_ = noveltyMapGenerator();

    {
        nav_msgs::msg::OccupancyGrid novelty_grid;
        novelty_grid.header = map->header;
        novelty_grid.info   = map->info;
        novelty_grid.data.resize(height * width);
        for (int r = 0; r < height; ++r)
            for (int c = 0; c < width; ++c)
                novelty_grid.data[r * width + c] =
                    static_cast<int8_t>(std::clamp(static_cast<int>(novelty_map_[r][c]), 0, 100));
        novelty_map_publisher_->publish(novelty_grid);
    }
}

// ---------------------------------------------------------------------------
// neighborsOf8
// ---------------------------------------------------------------------------
std::vector<std::pair<int,int>> NoveltyMapping::neighborsOf8(
    int row, int col,
    const std::vector<std::vector<int>>& map)
{
    const int rows = static_cast<int>(map_info_.height);
    const int cols = static_cast<int>(map_info_.width);

    const int dr[] = { 1, -1,  0,  0,  1, -1, -1,  1};
    const int dc[] = { 0,  0,  1, -1,  1, -1,  1, -1};

    std::vector<std::pair<int,int>> safe;
    for (int i = 0; i < 8; ++i) {
        int nr = row + dr[i];
        int nc = col + dc[i];
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && map[nr][nc] == 0)
            safe.emplace_back(nr, nc);
    }
    return safe;
}

// ---------------------------------------------------------------------------
// findFrontier
// ---------------------------------------------------------------------------
std::vector<std::vector<int>> NoveltyMapping::findFrontier(
    const std::vector<std::vector<int>>& map)
{
    const int rows = static_cast<int>(map_info_.height);
    const int cols = static_cast<int>(map_info_.width);

    std::vector<std::vector<int>> frontier(rows, std::vector<int>(cols, 0));

    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            if (map[r][c] == -1) {
                if (!neighborsOf8(r, c, map).empty())
                    frontier[r][c] = 100;
            }
        }
    }
    return frontier;
}

// ---------------------------------------------------------------------------
// noveltyMapGenerator
// ---------------------------------------------------------------------------
std::vector<std::vector<int8_t>> NoveltyMapping::noveltyMapGenerator()
{
    const int rows = static_cast<int>(map_info_.height);
    const int cols = static_cast<int>(map_info_.width);

    // Accumulate as int to avoid overflow before final clamp
    std::vector<std::vector<int>> accum(rows, std::vector<int>(cols, 0));

    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            if (frontier_map_[r][c] > 0) {
                auto visible = visibleCells(r, c);
                for (int vr = 0; vr < rows; ++vr)
                    for (int vc = 0; vc < cols; ++vc)
                        accum[vr][vc] += static_cast<int>(visible[vr][vc]);
            }
        }
    }

    std::vector<std::vector<int8_t>> novelty(rows, std::vector<int8_t>(cols, 0));
    for (int r = 0; r < rows; ++r)
        for (int c = 0; c < cols; ++c)
            novelty[r][c] = static_cast<int8_t>(std::clamp(accum[r][c], 0, 127));

    return novelty;
}

// ---------------------------------------------------------------------------
// visibleCells  (Bresenham-style ray casting)
// ---------------------------------------------------------------------------
std::vector<std::vector<int8_t>> NoveltyMapping::visibleCells(int ox, int oy)
{
    constexpr int EMPTY    =  0;
    constexpr int UNKNOWN  = 50;   // -1 raw → treated as unknown inside ray cast
    constexpr int OBSTACLE = 100;

    const int rows = static_cast<int>(map_info_.height);
    const int cols = static_cast<int>(map_info_.width);

    const double resolution    = map_info_.resolution;
    const int    max_range_cells = static_cast<int>(MAX_RANGE / resolution);

    if (ox < 0 || ox >= rows || oy < 0 || oy >= cols)
        throw std::out_of_range("Origin outside grid bounds");
    if (map_[ox][oy] == OBSTACLE)
        throw std::invalid_argument("Origin is inside an obstacle");

    std::vector<std::vector<int8_t>> visible(rows, std::vector<int8_t>(cols, 0));

    for (int i = 0; i < NUM_RAY_CASTS; ++i) {
        const double dx = cos_dirs_[i];
        const double dy = sin_dirs_[i];

        for (int step = 1; step <= max_range_cells; ++step) {
            int rx = static_cast<int>(std::round(ox + step * dx));
            int ry = static_cast<int>(std::round(oy + step * dy));

            if (rx < 0 || rx >= rows || ry < 0 || ry >= cols)
                break;  // Out of bounds

            int cell = map_[rx][ry];

            if (cell == OBSTACLE) {
                break;  // Blocked — stop ray, obstacle not counted
            } else if (cell == -1 && frontier_map_[rx][ry] < 50) {
                break;  // Unknown cell not on frontier — can't see past it
            } else if (cell == EMPTY) {
                visible[rx][ry] = 1;
            }
            // If it's a frontier cell (unknown but bordering free space), the
            // ray is NOT broken — matching the Python `elif cell == UNKNOWN and
            // frontier_map[rx,ry] < 50: break` logic (frontier cells pass through).
        }
    }

    return visible;
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------
int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<NoveltyMapping>();

    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);

    try {
        executor.spin();
    } catch (const std::exception & e) {
        RCLCPP_ERROR(node->get_logger(), "Exception: %s", e.what());
    }

    rclcpp::shutdown();
    return 0;
}

File: RBE-3002-D26-Team13-fast_pursuit/exploration/exploration/novelty_mapping.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node


from rclpy.executors import MultiThreadedExecutor
from  nav_msgs.msg import OccupancyGrid, MapMetaData
from rclpy.callback_groups import ReentrantCallbackGroup, MutuallyExclusiveCallbackGroup
import numpy as np


class NoveltyMapping(Node):

    def __init__(self):
        super().__init__('Novelty_Node')
        self.cb_group = MutuallyExclusiveCallbackGroup()

        self.subscriber_ = self.create_subscription(OccupancyGrid, '/map', self.map_callback, 10, callback_group=self.cb_group)

        self.novelty_map_publisher = self.create_publisher(OccupancyGrid, '/novelty', 10)

        self.frontier_map_publisher = self.create_publisher(OccupancyGrid, '/frontier', 10)

        self.map = np.zeros((0,0))
        self.map_info: MapMetaData = None

        self.frontier_map = np.zeros((0,0))
        self.frontier_map_info: MapMetaData = None

        self.novelty_map = np.zeros((0,0))
        self.novelty_map_info: MapMetaData = None
        self.NUM_RAY_CASTS = 18
        self.angles = np.linspace(0, 2 * np.pi, self.NUM_RAY_CASTS, endpoint=False)
        self.cos_dirs = np.cos(self.angles)
        self.sin_dirs = np.sin(self.angles)


    def map_callback(self, map):
        array_map = np.asarray(map.data, dtype=np.int32).reshape(map.info.height, map.info.width)

        # Store safe numpy array map to self.map and MapMetaData as member variables
        self.map =  array_map
        self.map_info = map.info

        self.frontier_map = self.find_frontier(array_map)
        self.frontier_map_info = map.info

        frontier_grid = OccupancyGrid()
        frontier_grid.header = map.header
        frontier_grid.info = map.info
        frontier_grid.data = self.frontier_map.flatten().astype(np.int32).tolist()

        self.frontier_map_publisher.publish(frontier_grid)

        novelty_map = self.novelty_map_generator()
        self.novelty_map = novelty_map
        self.novelty_map_info = map.info

        novelty_grid = OccupancyGrid()
        novelty_grid.header = map.header
        novelty_grid.info = map.info
        novelty_grid.data = np.clip(novelty_map, 0, 100).flatten().astype(np.int8).tolist()

        self.novelty_map_publisher.publish(novelty_grid)


    def neighbors_of_8(self, p: tuple[int, int], map) -> list[tuple[int, int]]:
        """
        Returns the safe 8-neighbors cells of (x,y) in the occupancy grid.
        :param p       [(int, int)]    The coordinate in the grid.
        :return        [[(int,int)]]   A list of walkable in 4 cardinal directions.
        """

        grid_x, grid_y = p

        # Create list of cells on all 4 sides of target
        neighbors = [
            (grid_x + 1, grid_y),
            (grid_x - 1, grid_y),
            (grid_x, grid_y + 1),
            (grid_x, grid_y - 1),
            (grid_x + 1, grid_y + 1),
            (grid_x - 1, grid_y - 1),
            (grid_x - 1, grid_y + 1),
            (grid_x + 1, grid_y - 1)
        ]

        safe_neighbors = []

        # Check each neighbor with the _is_safe function and add it to the list of safe neighbors
        for (x, y) in neighbors:
            if 0 <= x < self.map_info.height and 0 <= y < self.map_info.width:
                if map[x, y] == 0:
                    safe_neighbors.append((x, y))

        return safe_neighbors
    

    def find_frontier(self, map):
        unknown_cells = np.argwhere(map == -1)

        frontier_map = np.zeros(map.shape)
        


        for row, col in unknown_cells:
            if len(self.neighbors_of_8((row, col), map)) > 0:
                frontier_map[row, col] = 100

        return frontier_map


    def novelty_map_generator(self):

        novelty_grid = np.zeros(self.map.shape, dtype= np.int8)

        frontier_coords = np.argwhere(self.frontier_map)

        counter = 0

        divisor = int(len(frontier_coords)/50.0)

        if divisor == 0:
            divisor = 1

        for cord in frontier_coords:
            if counter % divisor == 0:
                visible_at_cord = self.visible_cells(cord)
                novelty_grid = novelty_grid + visible_at_cord

            counter += 1

        return novelty_grid


    def visible_cells(self, point: tuple[int, int]):
        MAX_RANGE = 1

        resolution = self.map_info.resolution
        max_range_cells = int(MAX_RANGE/resolution)
        
        
        EMPTY = 0
        UNKNOWN = 50
        OBSTACLE = 100

        grid = self.map

        rows, cols = grid.shape
        ox, oy = point

        if not (0 <= ox < rows and 0 <= oy < cols):
            raise ValueError(f"Origin {point} is outside grid bounds ({rows}, {cols})")
        if grid[ox, oy] == OBSTACLE:
            raise ValueError(f"Origin {point} is inside an obstacle")

        visible_map = np.zeros(grid.shape,dtype=np.int8)

        for i in range(self.NUM_RAY_CASTS):
            dx = self.cos_dirs[i]
            dy = self.sin_dirs[i]
            # Step along the ray using Bresenham-style sampling
            for step in range(1, max_range_cells + 1):
                rx = int(round(ox + step * dx))
                ry = int(round(oy + step * dy))

                # Stop if out of bounds
                if not (0 <= rx < rows and 0 <= ry < cols):
                    break

                cell = grid[rx, ry]

                if cell == OBSTACLE:
                    break  # Blocked — stop the ray, obstacle not counted
                elif cell == UNKNOWN and self.frontier_map[rx, ry] < 50:
                    break  # Can see the unknown cell but not past it
                elif cell == EMPTY:
                    visible_map[rx, ry] = 1

        return visible_map
    

def main(args=None):
    rclpy.init(args=args)

    novelty_mapping = NoveltyMapping()
    executor = MultiThreadedExecutor()
    executor.add_node(novelty_mapping)

    try:
        executor.spin()
    except KeyboardInterrupt:  # when 'ctrl + C' is pressed
        pass
    finally:
        novelty_mapping.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()